bittensor/axon.py

```diff
@@ -888,25 +890,38 @@ async def default_verify(self, synapse: bittensor.Synapse):
             # Build the unique endpoint key.
             endpoint_key = f"{synapse.dendrite.hotkey}:{synapse.dendrite.uuid}"
 
-            # Check the nonce from the endpoint key with 4 second delta
-            allowedDelta = 4000000000
-
             # Requests must have nonces to be safe from replays
             if synapse.dendrite.nonce is None:
                 raise Exception("Missing Nonce")
 
             # If we don't have a nonce stored, ensure that the nonce falls within
             # a reasonable delta.
+
+            # Updated nonce using NTP implementated at v7.2
             if (
-                self.nonces.get(endpoint_key) is None
-                and synapse.dendrite.nonce <= time.time_ns() - allowedDelta
-            ):
-                raise Exception("Nonce is too old")
-            if (
-                self.nonces.get(endpoint_key) is not None
-                and synapse.dendrite.nonce <= self.nonces[endpoint_key]
+                synapse.dendrite.version is not None
+                and synapse.dendrite.version >= V_7_2_0
             ):
-                raise Exception("Nonce is too old")
+                # If we don't have a nonce stored, ensure that the nonce falls within
+                # a reasonable delta.
+                current_time = BittensorNTPClient.get_current_ntp_time()
```

Comment:
it was accidental here, but if I would be malicious miner, this is the kind of thing I would try to push to official bittensor package so other miners are running suboptimal code

```
>>> client = ntplib.NTPClient()
>>> def ntp_request():
...     try: response = client.request('pool.ntp.org')
...     except: pass
... 
>>> timeit.timeit(ntp_request, number=10)
15.35390254703816
```
delay of 1.5s is going to tank your emissions in any subnet where they are affected by latency.

Sadly, it is against our (the company I work for) financial interests ( :money_with_wings: ) to even report this, but we try to be honest contributors to bittensor as a whole and do not put our interests before the Bittensor community.

Back to the technical part: This is also synchronous code, so this effect will be cumulative i.e. it will become very easy to DoS any server implementing this.

How this should be implemented:
My suggestion would be to save the development resources, and leave NTP synchronization to processes on Operating system side and use UTC time (as acquired from system clock) here instead.
This would mean a simple note in README about it.

Yes, they are configured a little bit differently (to drift towards synchronization instead of snapping into "right time" right away), but they do that job really well. 
Trying to reimplement NTP synchronization in will be yet another technical challenge consuming bittensor development team and contributors time of which there is always not enough to go around.

If there is enough time, then option to use in-process NTP synchronization could be added, but it should always be a opt-in / fallback, since we will never beat the efficiency of the NTP daemon process on OS side.

Reply:
Thanks for your valuable input. 

Previously, we also noticed some delays when hitting the `pool.ntp.org` directly - which is why we changed the endpoints to specific servers e.g: `0.pool.ntp.org` etc. This resulted in faster access times and was averaging 0.015 seconds per request. 

```
>>> timeit.timeit(ntp_request, number=10)
0.12425095890648663
>>> timeit.timeit(ntp_request, number=10)
0.10041512502357364
>>> timeit.timeit(ntp_request, number=10)
0.10449333302676678
>>> timeit.timeit(ntp_request, number=10)
0.13884595804847777
```

However, based on your results, I did a stress-test and found that some requests could take longer (around 1 - 1.5 seconds). Even though this was rare, but I am assuming it would surface more often in practice - considering real distributed network interactions as compared to us testing locally.

```
>>> timeit.timeit(ntp_request, number=10)
10.176145999925211
```

The main motivation to implement NTP was as you mentioned, to counter clock drifts, skews etc, all the issues with clocks in distributed systems. But at this point I would agree that reverting to system UNIX time would be a better approach. 

We can revisit this gap later on when we have cycles and see feedback regarding the new nonces.